File: Java/SpaceVIL/spacevil/src/main/java/com/spvessel/spacevil/CustomSelector.java
package com.spvessel.spacevil;

import java.awt.Point;
import java.util.LinkedList;
import java.util.List;

class CustomSelector extends Primitive {
    private static int count = 0;

    /**
     * Constructs a CustomSelector
     */
    public CustomSelector() {
        super("CustomSelector_" + count);
        count++;
    }

    /**
     * Make CustomSelector's rectangles with left top and right bottom points
     */
    public void setRectangles(List<Point> points) {
        List<float[]> triangles = new LinkedList<>();
        int w1 = 0, w2 = 0;
        int h1 = 0, h2 = 0;
        int rectX = 0, rectY = 0;
        if (points != null && points.size() > 0) {
            w1 = points.get(0).x;
            w2 = w1;
            h1 = points.get(0).y;
            h2 = h1;
            rectX = points.get(0).x;
            rectY = points.get(1).y;
            for (int i = 0; i < points.size() / 2; i++) {
                Point p1 = points.get(i * 2 + 0);
                Point p2 = points.get(i * 2 + 1);
                triangles.addAll(
                        GraphicsMathService.getRectangle((p2.x - p1.x), (p2.y - p1.y), p1.x - rectX, p2.y - rectY));
                w1 = (p1.x < w1) ? p1.x : w1;
                w2 = (p2.x > w2) ? p2.x : w2;
                h1 = (p1.y < h1) ? p1.y : h1;
                h2 = (p2.y > h2) ? p2.y : h2;
            }
        }
        setX(rectX);
        setY(rectY);
        setWidth(w2 - w1);
        setHeight(h2 - h1);
        setTriangles(triangles);
        ItemsRefreshManager.setRefreshShape(this);
    }

    /**
     * Shift selector on Y direction
     */
    public void shiftAreaY(int yShift) {
        List<float[]> triangles = getTriangles();
        if (triangles == null || triangles.size() == 0)
            return;

        for (float[] xyz : triangles) {
            xyz[1] += yShift;
        }

        setTriangles(triangles);
    }

    /**
     * Shift selector on X direction
     */
    public void shiftAreaX(int xShift) {
        List<float[]> triangles = getTriangles();
        if (triangles == null || triangles.size() == 0)
            return;

        for (float[] xyz : triangles) {
            xyz[0] += xShift;
        }

        setTriangles(triangles);
    }
}

File: CSharp/SpaceVIL/SpaceVIL/Items/CustomSelector.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Drawing;

namespace SpaceVIL
{
    internal class CustomSelector : Primitive
    {
        private static int count = 0;

        /// <summary>
        /// Constructs a CustomSelector
        /// </summary>
        internal CustomSelector() : base(name: "CustomSelector_" + count)
        {
            count++;
        }

        /// <summary>
        /// Make CustomSelector's rectangles with left top and right bottom points
        /// </summary>
        internal void SetRectangles(List<SpaceVIL.Core.Point> points)
        {
            List<float[]> triangles = new List<float[]>();
            int w1 = 0, w2 = 0;
            int h1 = 0, h2 = 0;
            int rectX = 0, rectY = 0;
            if (points != null && points.Count > 0)
            {
                w1 = points[0].X;
                w2 = w1;
                h1 = points[0].Y;
                h2 = h1;
                rectX = points[0].X;
                rectY = points[1].Y;
                for (int i = 0; i < points.Count / 2; i++)
                {
                    SpaceVIL.Core.Point p1 = points[i * 2 + 0];
                    SpaceVIL.Core.Point p2 = points[i * 2 + 1];
                    triangles.AddRange(
                        GraphicsMathService.GetRectangle((p2.X - p1.X), (p2.Y - p1.Y), p1.X - rectX, p2.Y - rectY));
                    w1 = (p1.X < w1) ? p1.X : w1;
                    w2 = (p2.X > w2) ? p2.X : w2;
                    h1 = (p1.Y < h1) ? p1.Y : h1;
                    h2 = (p2.Y > h2) ? p2.Y : h2;
                }
            }
            SetX(rectX);
            SetY(rectY);
            SetWidth(w2 - w1);
            SetHeight(h2 - h1);
            SetTriangles(triangles);
            ItemsRefreshManager.SetRefreshShape(this);
        }

        /// <summary>
        /// Shift selector on Y direction
        /// </summary>
        internal void ShiftAreaY(int yShift) {
            SetY(yShift);
            List<float[]> triangles = GetTriangles();
            if (triangles == null || triangles.Count == 0)
                return;
            
            foreach (float[] xyz in triangles) {
                xyz[1] += yShift;
            }
            SetTriangles(triangles);
        }

        /// <summary>
        /// Shift selector on X direction
        /// </summary>
        internal void ShiftAreaX(int xShift)
        {
            SetX(xShift);
            List<float[]> triangles = GetTriangles();
            if (triangles == null || triangles.Count == 0)
                return;

            foreach (float[] xyz in triangles)
            {
                xyz[0] += xShift;
            }
            SetTriangles(triangles);
        }
    }
}
